bin/setup_db

```diff
@@ -15,28 +15,21 @@
 # limitations under the License.
 # -----------------------------------------------------------------------------
 
-import os
-import sys
-import logging
 import argparse
+import logging
+import sys
+import os
+
 import rethinkdb as r
 from rethinkdb.errors import RqlRuntimeError
 
 LOGGER = logging.getLogger(__name__)
 LOGGER.level = logging.INFO
 LOGGER.addHandler(logging.StreamHandler(sys.stdout))
 
-
-def get_env_with_fallback(name, default):
-    value = os.getenv(name)
-    if value is None or value is "":
-        return default
-    return value
-
-
-DB_HOST = get_env_with_fallback('DB_HOST', 'localhost')
-DB_PORT = get_env_with_fallback('DB_PORT', '28015')
-DB_NAME = get_env_with_fallback('DB_NAME', 'rbac')
+DB_HOST = os.getenv('DB_HOST', 'localhost')
+DB_PORT = os.getenv('DB_PORT', '28015')
+DB_NAME = os.getenv('DB_NAME', 'rbac')
```

Comment:
I agree with your statement, but am confused about why this is a problem. 

Returning the blank string is the expected behavior of the function. That is: Assuming it is possible to create environment variables having blank strings, one would reason that people would be assigning them to blank strings when there is some purpose served in doing so. 

This function falls back to the default value when the key of the environment variable is absent (as it should). What you need to clarify is why environment variables having empty values would ever exist in the first place.

Reply:
Passing environment variables from .env through Docker Compose defaults them to blank strings.

Reply:
We do not wish to configure a default AES or Secret key in docker-compose, as it may encourage people to change it to their desired key values.

Reply:
Then I recommend we treat those two bits of data with special treatment rather than alter the behavior of os.getenv across the board because of them.

Reply:
The improvement was universal, blank strings are not useful values, but whatever.

Reply:
No, it wasn't. 

I covered your first comment in my original commit message, and the title reflects my position on how much of an improvement it was. 

The special case that came later (docker-compose plugging in blank values in the absence of environment variables). We have dummy values assigned in DEFAULT_CONFIG that are autogenerated when unchanged.

Also, you commented the same for (this comment) DB_NAME, so it is unclear why you have considered this effort a failure.

Finally, I have the branch up and running locally, opened the UI, and successfully executed the test_data_loader script. So what is this about an E2E test that would fail?

Reply:
You're right, the .env pass-throughs were not configured.
PR #572
That will expose the problem with blank strings on the AES key that the modified getenv function addressed.
It's required anytime a environment variable is passed through via docker compose without a default specified, for any variable. By convention we have decided to also supply defaults for most values in docker compose, however the problem is otherwise generic.
Your PR can't use only conventional getenv to apply defaults specifically for AES key and secret key, the others may so long as convention to specify defaults in docker compose is maintained.

Reply:
Is this a current problem, or a future problem? Because the file has been copied and pasted across the code base does not qualify it as a convention.

Reply:
I am marking this is unresolved because we need to sort out the matter of work that affects the current codebase and changes that affect something you perceive as being a problem in the future. Iterative and incremental development often means deleting or refactoring code in its most recent state.

It does not mean marking PRs as "Needs Work" over functionality that does not exist yet and then going so far as creating a new branch with the sole purpose of breaking a PR in progress. 

I understand the sensitivity of the AES key and secret, but supplying them to the application is another challenge altogether. I don't see how continually debating this minor change (less than one hour to implement, several hours debating, days lingering) accomplishes anything.

I would like to keep this conversation open for now because the matter is not resolved for me.

Reply:
You added to the codebase (PR #56) a default AES key generation with the goal of making the application work out of the box without requiring the user to create the .env file and generate their own key. Unfortunately that change you made was defective, because when you restart the application, you get a new AES key that can't decrypt the existing encrypted data (reproduce this bug using the persistent data flag).

We had to change our AES algorithm due to a security vulnerability in the old AES package, so I took that opportunity, given the old AES key generation was incompatible with the newly chosen AES library, to fix this defect. (PR #259)

https://github.com/hyperledger/sawtooth-next-directory/pull/259/files

Your reviewed, commented upon, and approved this pull request.

We know that the environment config sections of these daemons will get refactored soon, because we know we need to support secure secrets providers (TVAULT) for storing secrets. Issue #206. So it wasn't worth factoring out duplication now, given there is a pending feature /refactor of that code.

In PR #397 you changed the name in one file of getenv to get_env_with_fallback. I told you that that's not an appropriate name for the function because it doesn't describe how it differs from os.getenv. The difference is not that it has a default value (so does os.getenv), but rather that it treats blanks strings the same as no value and uses the default. This should be described in a docstring on the function. I also told you that same function exists in several places, due to PR #259, and should either be renamed consistently across the project or imported from a utility class. I approved your PR anyway.

Finally, in PR #533, you decide to remove the function rather than add the docstring and import it form a utility, despite the previous conversations about issue #206 and PR #397, and without asking the author (me) why the function is there.

I'm perfectly fine with breaking the application if users don't perform the manual step of creating a .env file and configuring a AES key; if your goal in PR #56 you no longer care about.

However, you can't break the Travis e2e tests that Ana is writing on the API, which will require a valid AES key without one manually configured. I'll fix that in another PR, and approve this one.

You need to resolve conflicts before merging.